### PR TITLE
Replace Jansson with Rust JsonBuilder - v27

### PR DESCRIPTION
**Pull request by jasonish**
Previous PR: https://github.com/OISF/suricata/pull/4901

Changes from last PR:
- convert http
- convert tls
- convert file/fileinfo
- convert anomaly

### CODE REVIEW

File: src/output-json.c [hunk 1]
```diff
@@ -694,6 +694,29 @@ void JsonTcpFlags(uint8_t flags, json_t *js)
...
+    if (flags & TH_FIN)
+        jb_set_bool(js, "fin", true);
+    if (flags & TH_RST)
+        jb_set_bool(js, "rst", true);
+    if (flags & TH_PUSH)
+        jb_set_bool(js, "psh", true);
+    if (flags & TH_ACK)
+        jb_set_bool(js, "ack", true);
+    if (flags & TH_URG)
+        jb_set_bool(js, "urg", true);
```
Comment (victorjulien) [inline src/output-json.c:713]: we could possibly micro-optimize calls like this as it produces a constant output `urg: true`

File: src/output-json.c [hunk 2]
```diff
@@ -694,6 +694,29 @@ void JsonTcpFlags(uint8_t flags, json_t *js)
         json_object_set_new(js, "cwr", json_true());
 }
 
+/** \brief jsonify tcp flags field
+ *  Only add 'true' fields in an attempt to keep things reasonably compact.
+ */
+void EveTcpFlags(uint8_t flags, JsonBuilder *js)
```
Comment (victorjulien) [inline src/output-json.c:700]: flags can be `const`

File: src/output-json.c (hunk 1 above)
Comment (victorjulien) [reply src/output-json.c:713]: In C at least, maybe a macro that just creates a constant string?

Untested:
```c
#define JB_SET_PREFORMATTED(jb, key, value) jb_add((jb), (key) ":" (value))
```

**Comment by victorjulien** [comment]
Seems passing NULL is a pattern we use after all.
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __strlen_sse2 () at ../sysdeps/x86_64/multiarch/../strlen.S:120
120     ../sysdeps/x86_64/multiarch/../strlen.S: No such file or directory.
[Current thread is 1 (Thread 0x7f195ffff700 (LWP 4508))]
(gdb) bt
#0  __strlen_sse2 () at ../sysdeps/x86_64/multiarch/../strlen.S:120
#1  0x0000555d44cdebfa in std::ffi::c_str::CStr::from_ptr ()
#2  0x0000555d44c0d19c in jb_set_string (js=0x7f19582443e0, key=<optimized out>, val=0x0) at src/jsonbuilder.rs:545
#3  0x0000555d44ae683d in AnomalyAppLayerDecoderEventJson (aft=0x7f19580f77f0, p=0x7f19f023afa0, decoder_events=0x7f1958837790, is_pktlayer=true, layer=0x555d44d5f1ee "proto_detect", tx_id=18446744073709551615) at output-json-anomaly.c:175
#4  0x0000555d44ae6bb3 in AnomalyJson (tv=0x555d467bc590, aft=0x7f19580f77f0, p=0x7f19f023afa0) at output-json-anomaly.c:253
#5  0x0000555d44ae6c64 in JsonAnomalyLogger (tv=0x555d467bc590, thread_data=0x7f19580f77f0, p=0x7f19f023afa0) at output-json-anomaly.c:274
#6  0x0000555d44b06ea8 in OutputPacketLog (tv=0x555d467bc590, p=0x7f19f023afa0, thread_data=0x7f19580e6700) at output-packet.c:116
#7  0x0000555d44ae1656 in OutputLoggerLog (tv=0x555d467bc590, p=0x7f19f023afa0, thread_data=0x7f19580e66e0) at output.c:933
#8  0x0000555d44acad6b in FlowWorker (tv=0x555d467bc590, p=0x7f19f023afa0, data=0x7f1958000b20) at flow-worker.c:289

(gdb) f 3
#3  0x0000555d44ae683d in AnomalyAppLayerDecoderEventJson (aft=0x7f19580f77f0, p=0x7f19f023afa0, decoder_events=0x7f1958837790, is_pktlayer=true, layer=0x555d44d5f1ee "proto_detect", tx_id=18446744073709551615) at output-json-anomaly.c:175
175             jb_set_string(js, "app_proto", alprotoname);
(gdb) p alprotoname
$1 = 0x0
```

**Comment by jasonish** [comment]
> Seems passing NULL is a pattern we use after all.
> 
> ```
> Program terminated with signal SIGSEGV, Segmentation fault.
> #0  __strlen_sse2 () at ../sysdeps/x86_64/multiarch/../strlen.S:120
> 120     ../sysdeps/x86_64/multiarch/../strlen.S: No such file or directory.
> [Current thread is 1 (Thread 0x7f195ffff700 (LWP 4508))]
> (gdb) bt
> #0  __strlen_sse2 () at ../sysdeps/x86_64/multiarch/../strlen.S:120
> #1  0x0000555d44cdebfa in std::ffi::c_str::CStr::from_ptr ()
> #2  0x0000555d44c0d19c in jb_set_string (js=0x7f19582443e0, key=<optimized out>, val=0x0) at src/jsonbuilder.rs:545
> #3  0x0000555d44ae683d in AnomalyAppLayerDecoderEventJson (aft=0x7f19580f77f0, p=0x7f19f023afa0, decoder_events=0x7f1958837790, is_pktlayer=true, layer=0x555d44d5f1ee "proto_detect", tx_id=18446744073709551615) at output-json-anomaly.c:175
> #4  0x0000555d44ae6bb3 in AnomalyJson (tv=0x555d467bc590, aft=0x7f19580f77f0, p=0x7f19f023afa0) at output-json-anomaly.c:253
> #5  0x0000555d44ae6c64 in JsonAnomalyLogger (tv=0x555d467bc590, thread_data=0x7f19580f77f0, p=0x7f19f023afa0) at output-json-anomaly.c:274
> #6  0x0000555d44b06ea8 in OutputPacketLog (tv=0x555d467bc590, p=0x7f19f023afa0, thread_data=0x7f19580e6700) at output-packet.c:116
> #7  0x0000555d44ae1656 in OutputLoggerLog (tv=0x555d467bc590, p=0x7f19f023afa0, thread_data=0x7f19580e66e0) at output.c:933
> #8  0x0000555d44acad6b in FlowWorker (tv=0x555d467bc590, p=0x7f19f023afa0, data=0x7f1958000b20) at flow-worker.c:289
> 
> (gdb) f 3
> #3  0x0000555d44ae683d in AnomalyAppLayerDecoderEventJson (aft=0x7f19580f77f0, p=0x7f19f023afa0, decoder_events=0x7f1958837790, is_pktlayer=true, layer=0x555d44d5f1ee "proto_detect", tx_id=18446744073709551615) at output-json-anomaly.c:175
> 175             jb_set_string(js, "app_proto", alprotoname);
> (gdb) p alprotoname
> $1 = 0x0
> ```

Resolved.

**Comment by jasonish** [comment]
Continued at https://github.com/OISF/suricata/pull/4929

File: src/output-json.c (hunk 1 above)
Comment (victorjulien) [reply src/output-json.c:713]: Any thoughts on this?

Comment (jasonish) [reply src/output-json.c:713]: This is what I've come up with that seems to work..  The Rust method is:
```
jb_set_formatted(JsonBuilder *js, const char *formatted);
```

To use with a bool like above from C:
```
jb_set_formatted(jb, "\"ack\":true");
```

I've also added a macro:
```
#define JB_SET_FORMATTED(jb, key, val) jb_set_formatted((jb), "\"" key "\":" val)
```
so the above would be:
```
JB_SET_FORMATTED(jb, "ack", "true");
```
however, to deal with https://github.com/OISF/suricata/pull/5030#discussion_r436290712, we'd have to do something like:
```
JB_SET_FORMATTED(jb, "command", "\"RETR\"");
```
but I set we could have JB_SET_STRING, JB_SET_TRUE, JB_SET_FALSE to handle some of these details.

Comment (jasonish) [reply src/output-json.c:713]: As a PR: https://github.com/OISF/suricata/pull/5042

**Pull request closed without merging**